### ISSUE TITLE
the performance of small object detection

### ISSUE THREAD

**Issue by jacobssy** [experiencor/basic-yolo-keras#191]
Hi,Thanks for sharing the code! it is really a great work you have done !
  I tried my own data followed by  the "Yolo Step_by_Step.ipynb", but the performance is not good.
the dataset is about traffic_signs including 39000 pictures divided to 43 classes.every picture is 60-120 pix, I Trained with 80 epoch, I adjust the input value from 416 to 192,the training index looks good, here is the screenshot:
![train](https://user-images.githubusercontent.com/14237914/37576374-0a39ebd6-2b67-11e8-9b57-1ced9f390b80.png)
but when I test the model in test data set(including 900 pictures,1360x800 pix,every picture includes 1-4 class of traffic_sign),it nearly not work, the detection boxes does not appear
So,I have a question about whether the yolo model could have good performance of small object detection,would you please give me some suggestions about my work? or some improvement about my training or testing ?

**Comment by jacobssy** [comment]
add a screenshot of  training index after 80 epoch:
![image](https://user-images.githubusercontent.com/14237914/37576688-7f1db6b6-2b68-11e8-8ebe-e205384776a7.png)

**Comment by Sayantan17** [comment]
@jacobssy buddy can u share the file with me as I am getting lots of errors. (Ipynb file)

**Comment by jacobssy** [comment]
@Sayantan17 Hi,I just modify some data input(according to my dataset),actually I did not do any other special work on the ipynb,I update the file to  my repo https://github.com/jacobssy/Traffic_Sign_detection/tree/master/detection , if you have some questions about the ipynb file ,I would try to help you :)

**Comment by Sayantan17** [comment]
@jacobssy  yeah sure.

**Comment by orsveri** [comment]
@jacobssy Hi, I had such problem. Have you tried to lower threshold of boxes confidence? In my case I set threshold to 0 and there was a boxes. Then I load weights to model, enlarged scales in config file (object_scale and class_scale) to 10, and made another 20-30 epochs. The result become better! Now I am trying different scale values and data to make result good for me.
![7](https://user-images.githubusercontent.com/26661572/37583033-9cc2a882-2b60-11e8-9816-28612bfecfd4.png)

**Comment by jacobssy** [comment]
@orsveri  oh , it looks good ,Thanks a lot !  I am trying that , do you also use the same dataset(GTSRB),it seems that you did not use so many classes of traffic signs, you just make all different traffic sign into one class to train the model.

**Comment by orsveri** [comment]
I am glad to help you :)
I used LISA dataset. I was afraid that large number of classes with tiny yolo will give me a bad result, so I started from one class. And there is quite different looking signs in my country, so now I am preparing my own dataset with several classes (about 10).

**Comment by jacobssy** [comment]
@orsveri  Hi, I have trained the yolo model again, this time ,I just modify the class of traffic sign  to one class ,I  use GTSDB data (including 900 pictures,1360X800 pix),and I tested in LISA dataset, I found it looks better than my last result.it seems to have no important relationship  with threshold ,scale value(the lower threshold ,the more objects will be detected ,refer to my screenshot),the result is worse than yours,you could also use the GTSDB dataset to test your model.
![high threshold](https://user-images.githubusercontent.com/14237914/37637830-e38f1bf8-2c44-11e8-8ada-e642d1330616.png)
![low threshold](https://user-images.githubusercontent.com/14237914/37637886-14ec2c36-2c45-11e8-915a-01b81488414a.png)

and I have a very critical question about multiple classes detection ..
  if I only have lots of small pictures of multiple traffic signs without annotation(more than 30000 pictures),and fewer big pictures of traffic signs with annotation ,what could I do to train my model?
and I think this question could be extended,in many cases, We also meet the same question, I tried some detection model,I found it has a very critical condition which should lots of data with annotation,but in fact ,we can not get lots of data with annotation in specific practical situation,so, have you some tips to do this kind work? if there any paper could be referred?
 sorry for my poor English,I tried to make my question understandable  :)

**Comment by orsveri** [comment]
@jacobssy Thank you, it's a good idea about testing on GTSDB, I will try when I have more time.
If you have only big pictures of signs (as I understand, with big signs) with annotation, and want to detect signs that have smaller size, you may use augmentation. I mean, you can zoom out your big-signs images, and fill borders with pixel values of edges. Also you can apply some geometric transformations to change image a little. It's a common way to make a dataset larger.

**Comment by andohuman** [comment]
I'm working on a similar data with straws to count and find the number of straws in the given image. However for some reason YOLO cannot seem to learn to predict straws. 
![img_20180219_102112_detected](https://user-images.githubusercontent.com/26820345/37864799-6382f256-2f99-11e8-8abd-7b1f40199f89.jpg)

Are there any possible solutions to improve my model ?

**Comment by experiencor** [comment]
@andohuman Yes. You may try to generate your own anchors using gen_anchors.py.

**Comment by andohuman** [comment]
@experiencor I am using my own anchors. I tried using 7 anchor boxes instead of five but there is very little difference.

**Comment by experiencor** [comment]
Ok. So I see lots of false negatives meaning that YOLO fails to detect many straws. Increasing object_class will definitely help in this case.

**Comment by jacobssy** [comment]
@R1234A Basically,I take the variable image size by experiments, I adjust the parameter from 416,448,480.. according to my gpu memory

**Comment by R1234A** [comment]
@jacobssy But u have 1360X800 pix dimension image. So if you are adjusting your parameters from 416,448,480.. , but that will make your image size smaller right and fixed too like (416X416) or (448X448) or (480X480). Can you tell to give exactly same image as input to Full Yolo architecture as 1360X800 pix. without fixing it??

**Comment by jacobssy** [comment]
@R1234A   emmm. here  code "image = cv2.resize(image, (self.config['IMAGE_H'], self.config['IMAGE_W'])" it will resize the image to the input parameter you set, if the size is 1360x800 pix, it will be changed to fit the input size when you use Fullyolo. sry my poor English, maybe I did not understand your question :)

**Comment by experiencor** [comment]
yes, that's what it does
